packages/gatsby-theme-docs/src/modules/sso/hooks/use-authentication.ts

```diff
@@ -0,0 +1,46 @@
+import { useContext, useEffect, useState } from 'react';
+import { useAuth0 } from '@auth0/auth0-react';
+import useSWR from 'swr';
+import { fetcherWithToken } from '../../self-learning/hooks/hooks.utils';
+import { useAuthToken } from '../../self-learning/hooks/use-auth-token';
+import ConfigContext, {
+  EFeatureFlag,
+  isFeatureEnabled,
+} from '../../../components/config-context';
+
+const useAuthentication = () => {
+  const { isAuthenticated: isAuth0Authenticated, ...rest } = useAuth0();
+  const [isAuthenticated, setIsAuthenticated] = useState(false);
+  const { learnApiBaseUrl, selfLearningFeatures } = useContext(ConfigContext);
+  const { getAuthToken } = useAuthToken();
+
+  const shouldFetchData =
+    isFeatureEnabled(EFeatureFlag.CourseStatus, selfLearningFeatures) &&
+    isAuth0Authenticated;
+
+  const apiEndpoint = `/api/courses`;
+
+  const { data, error, isLoading } = useSWR(
+    shouldFetchData ? apiEndpoint : null,
+    (url) => fetcherWithToken(url, getAuthToken, learnApiBaseUrl, 'GET'),
+    {
+      revalidateOnReconnect: false,
+      revalidateIfStale: false,
+      revalidateOnMount: true,
+      dedupingInterval: 3 * 60 * 1000, // allow revalidation each 3 minutes
+      revalidateOnFocus: true,
```

Comment:
Maybe document these choices in terms of what you want to achieve?  I assume they are specific to the situation inside the blocking hook and a tradeoff that tries to catch situations in which a user has no moodle token yet but is already logged in.   

by the way on that occasion: Did you consider making the logic more explicit in being conditional on whether a proxy API moodle token already exists?    The `user_session` cookie the API sets is HTTP only so it can't be read from JS but a common pattern seems to be that the API sets (and removes) another cookie that is JS readable but empty and only indicates the existence of the session but not the contents.

Reply:
> Maybe document these choices in terms of what you want to achieve? I assume they are specific to the situation inside the blocking hook and a tradeoff that tries to catch situations in which a user has no moodle token yet but is already logged in.

See above, I can add the comments inline in the code so it stays documented.

> 
> by the way on that occasion: Did you consider making the logic more explicit in being conditional on whether a proxy API moodle token already exists? The `user_session` cookie the API sets is HTTP only so it can't be read from JS but a common pattern seems to be that the API sets (and removes) another cookie that is JS readable but empty and only indicates the existence of the session but not the contents.

What if the cookie exists but it has another user information? Shall I perform a sort of client side pre-validation? It feels un-necessary, at least until we see a real performance problem.

Reply:
> What if the cookie exists but it has another user information? Shall I perform a sort of client side pre-validation? It feels un-necessary, at least until we see a real performance problem. 

The idea was meant only for the case you think it's worth the effort at all.  But it would not have to handle edge cases because the purpose is performance, not precision  (there are other scenarios remaining where parallel requests can have no session yet, e.g. when the moodle token expired in the middle of a session).   

I don't think explicitly handling whether there is a session or not is giving a relevant server load gain because the SWR based approach would nevertheless "ping" the courses list every few minutes.  The advantage would be that the implementation would make the purpose of why it exists very explicit which may be good in the long run.  

What I'm not sure about is the impact on the visual loading experience though if it's done without SWR caching across page loads and without checking the session.   For example, navigating to a page with login / logout specific UI elements would have the log in / out specific elements have to "wait" until the courses are loaded even if there is no learning specific information on that page.    I can't really tell you exactly, but my gut feeling says that the feature would be cleaner if it explicitly knew whether there's a session and would then explicitly do nothing but calling a specialized API that does nothing but initializing the session if there is none yet (no SWR etc, directly calling it).